File: cpp/Graph/1514.path-with-maximum-probability.cpp
#include "../std.hpp"
class Solution {
public:
    double maxProbability(int n, vector<vector<int>> &edges, vector<double> &succProb, int start, int end)
    {
        vector<pair<int, double>> adj[n];
        for (int i = 0; i < edges.size(); i++) {
            int src = edges[i][0];
            int dest = edges[i][1];
            adj[src].push_back({dest, succProb[i]});
            adj[dest].push_back({src, succProb[i]});
        }
        priority_queue<pair<double, int>> pq;
        vector<double> dist(n, 0);
        pq.push({1, start});
        dist[start] = 1;
        while (!pq.empty()) {
            int u = pq.top().second;
            pq.pop();
            for (auto [v, prob] : adj[u]) {
                if (dist[v] < dist[u] * prob) {
                    dist[v] = dist[u] * prob;
                    pq.push({dist[v], v});
                }
            }
        }
        return dist[end];
    }
};
int main()
{
    Solution s;
    vector<vector<int>> edge = {{0, 1}, {1, 2}, {0, 2}};
    vector<double> succProb = {0.5, 0.5, 0.2};
    s.maxProbability(3, edge, succProb, 0, 2);
}

File: js/Graph/1514.path-with-maximum-probability.ts

//bellman ford(TLE)
function maxProbabilityBF(
  n: number,
  edges: number[][],
  succProb: number[],
  start: number,
  end: number
): number {
  let dist = new Array<number>(n).fill(0)
  dist[start] = 1
  for (let i = 1; i < n; ++i) {
    for (let j = 0; j < edges.length; ++j) {
      let [src, dest] = edges[j]
      dist[dest] = Math.max(dist[dest], dist[src] * succProb[j])
      dist[src] = Math.max(dist[src], dist[dest] * succProb[j])
    }
  }
  return dist[end]
}

//floyd warshall(TLE)
function maxProbabilityFW(
  n: number,
  edges: number[][],
  succProb: number[],
  start: number,
  end: number
): number {
  let dist = Array.from({ length: n }, () => Array<number>(n).fill(0))
  for (let i = 0; i < edges.length; ++i) {
    let [src, dest] = edges[i]
    dist[src][dest] = succProb[i]
    dist[dest][src] = succProb[i]
  }
  for (let i = 0; i < n; ++i) {
    dist[i][i] = 1
  }
  for (let k = 0; k < n; ++k) {
    for (let i = 0; i < n; ++i) {
      for (let j = 0; j < n; ++j) {
        dist[i][j] = Math.max(dist[i][j], dist[i][k] * dist[k][j])
      }
    }
  }
  return dist[start][end]
}
console.log(
  maxProbabilityBF(
    5,
    [
      [1, 4],
      [2, 4],
      [0, 4],
      [0, 3],
      [0, 2],
      [2, 3],
    ],
    [0.37, 0.17, 0.93, 0.23, 0.39, 0.04],
    3,
    4
  )
)
